src/test/java/uk/nhs/tis/trainee/usermanagement/service/MetricsServiceTest.java

```diff
@@ -0,0 +1,131 @@
+ * including without limitation the rights to use, copy, modify, merge, publish, distribute,
+ * sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
+ * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.nhs.tis.trainee.usermanagement.service;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static uk.nhs.tis.trainee.usermanagement.service.MetricsService.METRIC_NAME_ACCOUNT_DELETE;
+import static uk.nhs.tis.trainee.usermanagement.service.MetricsService.METRIC_NAME_MFA_RESET;
+import static uk.nhs.tis.trainee.usermanagement.service.MetricsService.METRIC_RESYNC;
+import static uk.nhs.tis.trainee.usermanagement.service.MetricsService.TAG_MFA;
+import static uk.nhs.tis.trainee.usermanagement.service.MetricsService.TAG_USER_STATUS;
+
+import com.amazonaws.services.cognitoidp.model.UserStatusType;
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.instrument.cumulative.CumulativeCounter;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.EnumSource;
+import org.junit.jupiter.params.provider.MethodSource;
+import uk.nhs.tis.trainee.usermanagement.enumeration.MfaType;
+
+class MetricsServiceTest {
+
+  private MetricsService service;
+
+  @BeforeEach
+  void setup() {
+    MeterRegistry meterRegistry = mock(MeterRegistry.class);
+    Meter.Id idAccountDelete = new Meter.Id("delete", Tags.empty(), null, null, Meter.Type.COUNTER);
+    Counter counterAccountDelete = new CumulativeCounter(idAccountDelete);
+
+    when(meterRegistry.counter(eq(METRIC_NAME_ACCOUNT_DELETE),
+        any(), any(),
+        eq(TAG_USER_STATUS), any(),
+        eq(TAG_MFA), any())).thenReturn(counterAccountDelete);
```

Comment:
This does not set up different counters for each permutation of tags as in the service itself, simply a single counter. As such, the tests consequently do not check that the correct specific counter is being used, only that the correct general type of counter (account-delete, mfa-reset, sync) is being used.

Reply:
(I did this because the tests felt clumsy enough as it is)

Reply:
From what I remember I had a nightmare writing tests for the metrics too, half the AWS library objects can't be easily constructed or mocked due to how they're initialised.